<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <arg name="namespace" default="/"/>
  <arg name="rviz"  default="true"/>

  <arg name="arm"  default="True"/>
  <arg name="ft_sensor"  default="schunk-ft"/>
  <arg name="laser_model"  default="sick-571"/>
  <arg name="camera_model"  default="orbbec-astra"/>
  <arg name="robot"      default="steel"/>
  <arg name="public_sim" default="true"/>
  <arg name="gzclient" default="true"/>
  <arg name="recording" default="false"/>
  <arg name="extra_gazebo_args" default=""/>
  <arg name="world"  default="tracking_aruco"/>
  <arg name="gzpose" default="-x 0.0 -y 0.0 -z 0.0 -R 0.0 -P 0.0 -Y 0.0"/>
  <arg name="tuck_arm"  default="true"/>

  <arg name="publish_all_markers" default="true"/> <!--decide to publish all the marker ids or only a certain "markerId"-->
  <arg name="markerId"        default="582"/>
  <arg name="markerSize"      default="0.16"/>    <!-- in m -->	
  <arg name="marker_frame"    default="aruco_marker_frame"/>
  <arg name="ref_frame"       default="odom"/>  <!-- leave empty and the pose will be published wrt param parent_name -->

  <arg name="planner"        default="base"/>
  <arg name="global_planner" default="global_planner"/>
  <arg name="local_planner"  default="$(eval 'teb' if arg('public_sim') else 'pal')"/>
  <arg name="localization"   default="amcl"/>
  <arg name="map"            default="$(env HOME)/.pal/tiago_maps/configurations/$(arg world)/config"/>
  <arg name="advanced_navigation"   default="false"/> <!-- Requires extra software from PAL Robotics-->
  <arg name="mapping"        default="$(eval 'gmapping' if arg('public_sim') else 'karto')"/>
  
  <arg name="sim_sufix" value="_public_sim"     if="$(arg public_sim)"/>
  <arg name="sim_sufix" value=""                unless="$(arg public_sim)"/>
  
  <arg name="crop_to_sort" default="Lemon"/>  
  
  <env name="LC_NUMERIC" value="C" />

 <group ns="$(arg namespace)">

  <!-- Gazebo simulation -->
  <include file="$(find tiago_gazebo)/launch/tiago_gazebo.launch">
    <arg name="robot"      value="$(arg robot)"/>
    <arg name="public_sim" value="$(arg public_sim)"/>
    <arg name="world"      value="$(arg world)"/>
    <arg name="gui"      value="$(arg gzclient)"/>
    <arg name="arm" value="$(arg arm)"/>
    <arg name="ft_sensor" value="$(arg ft_sensor)"/>
    <arg name="laser_model" value="$(arg laser_model)"/>
    <arg name="camera_model" value="$(arg camera_model)"/>
    <arg name="gzpose"     value="$(arg gzpose)"/>
    <arg name="tuck_arm"   value="$(arg tuck_arm)"/>
    <arg name="recording" value="$(arg recording)"/>
    <arg name="extra_gazebo_args" value="$(arg extra_gazebo_args)"/>
  </include>

   <!-- marker detector -->
   <node pkg="aruco_ros" type="single"  name="aruco_single">
       <remap from="/camera_info"       to="/xtion/rgb/camera_info" />
       <remap from="/image"             to="/xtion/rgb/image_rect_color" />
       <param name="image_is_rectified" value="True"/>
       <param name="marker_size"        value="$(arg markerSize)"/>
       <param name="marker_id"          value="$(arg markerId)"/>
       <param name="reference_frame"    value="$(arg ref_frame)"/>   <!-- frame in which the marker pose will be expressed -->
       <param name="publish_all_markers" value="$(arg publish_all_markers)"/>
       <param name="camera_frame"       value="xtion_rgb_optical_frame"/>
       <param name="marker_frame"       value="$(arg marker_frame)" />
       <param name="corner_refinement"  value="SUBPIX" />
   </node>

   <!-- navigation and mapping -->
   <include file="$(find tiago_2dnav_gazebo)/launch/navigation.launch">
      <arg name="state"          value="mapping"/>
      <arg name="mapping"        value="$(arg mapping)"/> <!--the need only for mapping-->
      <arg name="planner"        value="$(arg planner)"/>
      <arg name="global_planner" value="$(arg global_planner)"/>
      <arg name="local_planner"  value="$(arg local_planner)"/>
      <arg name="localization"   value="$(arg localization)"/>
      <arg name="map"            value="$(arg map)"/>
      <arg name="public_sim"     value="$(arg public_sim)"/>
      <arg name="rgbd_sensors"   value="$(arg advanced_navigation)"/>
      <arg name="laser_model"    value="$(arg laser_model)"/>
   </include>

   <group if="$(arg advanced_navigation)">
      <node name="rviz" pkg="rviz" type="rviz" if="$(arg rviz)"
            args="-d $(find tiago_2dnav)/config/rviz/advanced_navigation.rviz"/>
   </group>
      <group unless="$(arg advanced_navigation)">
       <node name="rviz" pkg="rviz" type="rviz" if="$(arg rviz)"
            args="-d $(find tiago_2dnav)/config/rviz/navigation$(arg sim_sufix).rviz"/>
      </group>
   </group>

   <!-- pointcloud to laserscan-->
  <group if="$(arg public_sim)">
    <include file="$(find tiago_laser_sensors)/launch/rgbd_cloud_laser.launch">
      <arg name="cloud" value="/xtion/depth_registered/points"/>
    </include>
  </group>

  <group if="$(arg advanced_navigation)">
    <!-- Advanced navigation -->
    <include file="$(find pal_head_manager)/launch/tiago_head_manager.launch"/>
    <node pkg="tiago_2dnav" type="navigation_camera_mgr.py" name="navigation_camera_mgr" />
    <node pkg="pal_zoi_detector" type="zoi_detector_node" name="zoi_detector" />
    <include file="$(find pal_waypoint)/launch/pal_waypoint.launch"/>
    <include file="$(find pal_navigation_manager)/launch/poi_navigation_server.launch"/>
    <include file="$(find tiago_laser_sensors)/launch/rgbd_cloud_laser.launch">
      <arg name="cloud"  value="/xtion/depth_registered/points"/>
    </include>
    <include file="$(find pal_map_utils)/launch/map_configuration_server.launch">
      <env name="PAL_HOST" value="tiagoc"/>
    </include>
  </group>

   <!-- Robot goal pose-->
   <node pkg="tracking_aruco" type="move_robot_base.py" name="move_robot_base_client">
  </node>

  <!-- Aruco markers navigation-->
   <!--node pkg="tracking_aruco" type="navigate_with_aruco.py" name="navigate_with_aruco" args="crop_to_sort">
    <param  name="crop_to_sort"  value="$(arg crop_to_sort)"/>
  </node-->
  
</launch>
